File: ai/services/aiClient.js
/**
 * AI Client — Dual provider
 * PRIMARY: Ollama (local, no quota)
 * FALLBACK: Gemini API (khi Ollama không có sẵn)
 * 
 * Theo request_ai.md: "Ollama hoặc thuật toán gợi ý"
 */
const { GoogleGenerativeAI } = require('@google/generative-ai');
const env = require('../../backend/src/config/env');

// ============ OLLAMA (PRIMARY) ============

const OLLAMA_BASE = env.OLLAMA_BASE_URL || 'http://localhost:11434';
const OLLAMA_MODEL = env.OLLAMA_MODEL || 'llama3.1';

/**
 * Check Ollama có sẵn không
 */
async function isOllamaAvailable() {
  try {
    const res = await fetch(`${OLLAMA_BASE}/api/tags`, { 
      signal: AbortSignal.timeout(3000) 
    });
    if (!res.ok) return false;
    const data = await res.json();
    // Kiểm tra model requested có trong list không
    const models = data.models || [];
    return models.some(m => m.name.includes(OLLAMA_MODEL.split(':')[0]));
  } catch {
    return false;
  }
}

/**
 * Gọi Ollama (non-streaming)
 */
async function callOllama(prompt, { system = '', temperature = 0.7 } = {}) {
  const messages = [];
  if (system) messages.push({ role: 'system', content: system });
  messages.push({ role: 'user', content: prompt });

  const res = await fetch(`${OLLAMA_BASE}/api/chat`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      model: OLLAMA_MODEL,
      messages,
      stream: false,
      options: { temperature, num_predict: 4096 },
    }),
    signal: AbortSignal.timeout(120000), // 2 phút timeout
  });

  if (!res.ok) {
    const err = await res.text().catch(() => res.status);
    throw new Error(`Ollama error ${res.status}: ${err}`);
  }

  const data = await res.json();
  return data.message?.content || data.response || '';
}

/**
 * Gọi Ollama streaming (chat format)
 */
async function* streamOllama(messages, { temperature = 0.7 } = {}) {
  const res = await fetch(`${OLLAMA_BASE}/api/chat`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      model: OLLAMA_MODEL,
      messages,
      stream: true,
      options: { temperature, num_predict: 2048 },
    }),
    signal: AbortSignal.timeout(120000),
  });

  if (!res.ok) {
    const err = await res.text().catch(() => res.status);
    throw new Error(`Ollama stream error ${res.status}: ${err}`);
  }

  const reader = res.body;
  const decoder = new TextDecoder();
  for await (const chunk of reader) {
    const lines = decoder.decode(chunk, { stream: true }).split('\n').filter(Boolean);
    for (const line of lines) {
      try {
        const data = JSON.parse(line);
        const text = data.message?.content || '';
        if (text) yield text;
      } catch { /* skip */ }
    }
  }
}

// ============ GEMINI (FALLBACK) ============

let genAI = null;
function getGemini() {
  if (!genAI && env.GEMINI_API_KEY) {
    genAI = new GoogleGenerativeAI(env.GEMINI_API_KEY);
  }
  return genAI;
}

async function callGemini(prompt, { systemInstruction = '', temperature = 0.7, maxTokens = 4096 } = {}) {
  const ai = getGemini();
  if (!ai) throw new Error('GEMINI_API_KEY not configured');

  const model = ai.getGenerativeModel({
    model: 'gemini-2.0-flash',
    systemInstruction: systemInstruction || undefined,
    generationConfig: { temperature, maxOutputTokens: maxTokens },
  });

  const result = await model.generateContent(prompt);
  return result.response.text();
}

async function* streamGeminiChat(systemPrompt, history, userMessage, { temperature = 0.7 } = {}) {
  const ai = getGemini();
  if (!ai) throw new Error('GEMINI_API_KEY not configured');

  const model = ai.getGenerativeModel({
    model: 'gemini-2.0-flash',
    systemInstruction: systemPrompt || undefined,
    generationConfig: { temperature, maxOutputTokens: 4096 },
  });

  const geminiHistory = history.map(msg => ({
    role: msg.role === 'assistant' ? 'model' : 'user',
    parts: [{ text: msg.content }],
  }));

  const chat = model.startChat({ history: geminiHistory });
  const result = await chat.sendMessageStream(userMessage);
  for await (const chunk of result.stream) {
    const text = chunk.text();
    if (text) yield text;
  }
}

// ============ SMART ROUTER ============

/**
 * generateText — Ollama primary, Gemini fallback
 * Dùng cho: roadmap generation, job suggestions (non-streaming)
 */
async function generateText(prompt, options = {}) {
  const ollamaOk = await isOllamaAvailable();

  if (ollamaOk) {
    try {
      console.log(`🦙 Calling Ollama (${OLLAMA_MODEL})...`);
      const result = await callOllama(prompt, {
        system: options.systemInstruction || '',
        temperature: options.temperature ?? 0.7,
      });
      console.log('✅ Ollama response OK');
      return result;
    } catch (err) {
      console.warn('⚠️ Ollama failed:', err.message);
    }
  } else {
    console.log('🦙 Ollama not available, trying Gemini...');
  }

  // Gemini fallback
  if (env.GEMINI_API_KEY) {
    try {
      console.log('✨ Calling Gemini...');
      const result = await callGemini(prompt, options);
      console.log('✅ Gemini response OK');
      return result;
    } catch (err) {
      console.warn('⚠️ Gemini failed:', err.message);
      throw new Error(`AI không khả dụng. Ollama: không kết nối được. Gemini: ${err.message.substring(0, 100)}`);
    }
  }

  throw new Error('Không có AI provider nào được cấu hình. Hãy chạy Ollama hoặc thêm GEMINI_API_KEY.');
}

/**
 * chatStream — Ollama primary, Gemini fallback
 * Dùng cho: chatbot streaming
 */
async function* chatStream(history, userMessage, options = {}) {
  const systemPrompt = options.systemInstruction || '';
  const ollamaOk = await isOllamaAvailable();

  if (ollamaOk) {
    try {
      console.log(`🦙 Chatting with Ollama (${OLLAMA_MODEL})...`);
      // Build messages array cho Ollama chat format
      const messages = [];
      if (systemPrompt) messages.push({ role: 'system', content: systemPrompt });
      for (const msg of history) {
        messages.push({
          role: msg.role === 'assistant' ? 'assistant' : 'user',
          content: msg.content,
        });
      }
      messages.push({ role: 'user', content: userMessage });

      let hasYielded = false;
      const gen = streamOllama(messages, { temperature: options.temperature ?? 0.7 });
      for await (const chunk of gen) {
        hasYielded = true;
        yield chunk;
      }
      if (hasYielded) return;
    } catch (err) {
      console.warn('⚠️ Ollama chat failed:', err.message);
    }
  }

  // Gemini fallback
  if (env.GEMINI_API_KEY) {
    try {
      console.log('✨ Chatting with Gemini...');
      const gen = streamGeminiChat(systemPrompt, history, userMessage, { temperature: options.temperature ?? 0.7 });
      for await (const chunk of gen) {
        yield chunk;
      }
      return;
    } catch (err) {
      console.warn('⚠️ Gemini chat failed:', err.message);
    }
  }

  throw new Error('AI không khả dụng. Vui lòng chạy: ollama pull ' + OLLAMA_MODEL);
}

// generateStream — alias của chatStream cho backwards compat
async function* generateStream(prompt, options = {}) {
  const messages = [];
  if (options.systemInstruction) messages.push({ role: 'system', content: options.systemInstruction });
  messages.push({ role: 'user', content: prompt });
  yield* streamOllama(messages, { temperature: options.temperature });
}

module.exports = { generateText, generateStream, chatStream, isOllamaAvailable };

File: ai/services/ai_client.py
"""
AI Client — OpenRouter
1 API key, chon model trong request.
Chat dung model nhanh, analysis dung model manh.
"""
from openai import OpenAI
from config import OPENROUTER_API_KEY, OPENROUTER_BASE_URL, CHAT_MODEL, ANALYSIS_MODEL
import httpx

DEFAULT_HEADERS = {
    "HTTP-Referer": "http://localhost:5173",
    "X-OpenRouter-Title": "EduPath AI",
}

# 1 client duy nhat
client = OpenAI(
    base_url=OPENROUTER_BASE_URL,
    api_key=OPENROUTER_API_KEY,
    http_client=httpx.Client(timeout=httpx.Timeout(180.0, connect=15.0)),
    default_headers=DEFAULT_HEADERS,
)


def generate_text(system_prompt: str, user_prompt: str, temperature: float = 0.7, model: str = None) -> str:
    """Non-streaming cho roadmap/jobs. Fallback: analysis -> chat model."""
    models_to_try = [model or ANALYSIS_MODEL, CHAT_MODEL]
    # Bo trung lap
    seen = set()
    models_to_try = [m for m in models_to_try if m not in seen and not seen.add(m)]

    last_error = None
    for m in models_to_try:
        try:
            response = client.chat.completions.create(
                model=m,
                messages=[
                    {"role": "system", "content": system_prompt},
                    {"role": "user", "content": user_prompt},
                ],
                temperature=temperature,
                max_tokens=3000,
            )
            result = response.choices[0].message.content or ""
            if result.strip():
                print(f"[OK] generate_text: {m}")
                return result
        except Exception as e:
            print(f"[WARN] Model {m} failed: {e}")
            last_error = e
            continue

    raise last_error or Exception("All models failed")


def generate_stream(system_prompt: str, messages: list, temperature: float = 0.7, model: str = None):
    """Streaming cho chatbot. Dung chat model (nhanh)."""
    target_model = model or CHAT_MODEL
    full_messages = [{"role": "system", "content": system_prompt}] + messages

    try:
        stream = client.chat.completions.create(
            model=target_model,
            messages=full_messages,
            temperature=temperature,
            max_tokens=800,  # Giảm từ 1024 → TTFT nhanh hơn
            stream=True,
        )
        print(f"[OK] generate_stream: {target_model}")
        for chunk in stream:
            if chunk.choices and chunk.choices[0].delta.content:
                yield chunk.choices[0].delta.content
    except Exception as e:
        print(f"[WARN] {target_model} failed: {e}, trying {ANALYSIS_MODEL}...")
        stream = client.chat.completions.create(
            model=ANALYSIS_MODEL,
            messages=full_messages,
            temperature=temperature,
            max_tokens=1024,
            stream=True,
        )
        for chunk in stream:
            if chunk.choices and chunk.choices[0].delta.content:
                yield chunk.choices[0].delta.content
